rpc_util_test.go

```diff
@@ -290,3 +293,138 @@ func BenchmarkGZIPCompressor512KiB(b *testing.B) {
+			error:                 errors.New("decompression error"),
+		},
+		{
+			name:       "Fails with overflow error when decompressed size exceeds maxReceiveMessageSize",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("overflow test data"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: 5,
+			want:                  nil,
+			compressedSize:        6, //len([]byte("overflow test data")),
+			error:                 errors.New("overflow: received message size is larger than the allowed maxReceiveMessageSize (5 bytes)."),
+		},
+		{
+			name:       "Returns EOF error when maxReceiveMessageSize is MaxInt64 but data ends prematurely",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("small data"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: math.MaxInt64,
+			want: func() mem.BufferSlice {
+				smallDecompressed := []byte("small data")
+				return mem.BufferSlice{mem.NewBuffer(&smallDecompressed, nil)}
+			}(),
+			compressedSize: len([]byte("small data")),
+			error:          nil,
+		},
+		{
+			name:                  "Fails with io.Copy error during decompression",
+			compressor:            c,
+			input:                 mem.BufferSlice{},
+			maxReceiveMessageSize: 100,
+			want:                  nil,
+			compressedSize:        0,
+			error:                 errors.New("simulated io.Copy read error"),
+		},
+		{
+			name:       "Succeeds when message size is much smaller than maxReceiveMessageSize",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("small message"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: 50,
+			want: func() mem.BufferSlice {
+				decompressed := []byte("small message")
+				return mem.BufferSlice{mem.NewBuffer(&decompressed, nil)}
+			}(),
+			compressedSize: len([]byte("small message")),
+			error:          nil,
+		},
+		{
+			name:       "Succeeds with MaxInt64 receive size and small decompressed data",
+			compressor: c,
+			input: func() mem.BufferSlice {
+				compressedData := compressData([]byte("small data"))
+				return mem.BufferSlice{mem.NewBuffer(&compressedData, nil)}
+			}(),
+			maxReceiveMessageSize: math.MaxInt64,
+			want: func() mem.BufferSlice {
+				decompressed := []byte("small data")
+				return mem.BufferSlice{mem.NewBuffer(&decompressed, nil)}
+			}(),
+			compressedSize: len([]byte("small data")),
+			error:          nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			output, size, err := decompress(tt.compressor, tt.input, tt.maxReceiveMessageSize, nil)
+			// Check if both err and tt.error are either nil or non-nil
+			if (err != nil) != (tt.error != nil) {
+				t.Errorf("decompress() error, got err=%v, want err=%v", err, tt.error)
```

Comment:
all should change to t.Fatalf because we don't want to continue once detecting an error. Like you won't compare compressed size if decompression has returned error